components/casemap/src/greek_to_me/mod.rs

```diff
@@ -12,131 +12,357 @@
+    // 0 is purposely left out so that the all-zero case is unambiguous
+    Α = 1,
+    Ε = 2,
+    Η = 3,
+    Ι = 4,
+    Ο = 5,
+    Υ = 6,
+    Ω = 7,
+
+    // This does not currently take any diacritics, but is technically
+    // still a vowel
+    ϒ = 8,
 }
 
-impl GreekPrecomposedLetterData {
-    /// Data on the diacritics present in this character
-    pub(crate) fn diacritics(self) -> GreekDiacritics {
-        GreekDiacritics {
-            accented: self.accented,
-            dialytika: self.dialytika,
-            ypogegrammeni: self.ypogegrammeni,
+impl From<GreekVowel> for char {
+    fn from(other: GreekVowel) -> Self {
+        match other {
+            GreekVowel::Α => 'Α',
+            GreekVowel::Ε => 'Ε',
+            GreekVowel::Η => 'Η',
+            GreekVowel::Ι => 'Ι',
+            GreekVowel::Ο => 'Ο',
+            GreekVowel::Υ => 'Υ',
+            GreekVowel::Ω => 'Ω',
+            GreekVowel::ϒ => 'ϒ',
         }
     }
-    /// The base greek letter in this character, uppercased (if any)
-    pub(crate) fn greek_base_uppercase(self) -> Option<char> {
-        self.uppercase
+}
+
+impl TryFrom<char> for GreekVowel {
+    type Error = ();
+    fn try_from(other: char) -> Result<Self, ()> {
+        Ok(match other {
+            'Α' => GreekVowel::Α,
+            'Ε' => GreekVowel::Ε,
+            'Η' => GreekVowel::Η,
+            'Ι' => GreekVowel::Ι,
+            'Ο' => GreekVowel::Ο,
+            'Υ' => GreekVowel::Υ,
+            'Ω' => GreekVowel::Ω,
+            'ϒ' => GreekVowel::ϒ,
+            _ => return Err(()),
+        })
     }
+}
 
-    /// Is this a greek letter?
-    pub(crate) fn is_greek_letter(self) -> bool {
-        self.uppercase.is_some()
+impl TryFrom<u8> for GreekVowel {
+    type Error = ();
+    fn try_from(other: u8) -> Result<Self, ()> {
+        Ok(match other {
+            1 => Self::Α,
+            2 => Self::Ε,
+            3 => Self::Η,
+            4 => Self::Ι,
+            5 => Self::Ο,
+            6 => Self::Υ,
+            7 => Self::Ω,
+            8 => Self::ϒ,
+            _ => return Err(()),
+        })
     }
 }
 
-impl From<u16> for GreekPrecomposedLetterData {
-    fn from(sixteen: u16) -> Self {
-        let mut this = Self::default();
-        if sixteen & (1 << 8) != 0 {
-            let int = u32::from(sixteen & 0xFF);
-            let ch = int + GREEK_BLOCK_START;
-            let ch = char::try_from(ch);
-            debug_assert!(ch.is_ok());
-            this.uppercase = ch.ok();
-        } else {
-            debug_assert!((sixteen & 0xFF00) == 0)
-        }
-        this.ypogegrammeni = sixteen & (1 << 9) != 0;
-        this.dialytika = sixteen & (1 << 10) != 0;
-        this.accented = sixteen & (1 << 11) != 0;
-        debug_assert!(
-            sixteen & (0xFFFF - (1 << 12) + 1) == 0,
-            "Bits 12..16 should not be set in GreekPrecomposedLetterData"
-        );
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum GreekConsonant {
```

Comment:
Ideally we would be able to store the consonant as an offset from 0x370. This works for all base consonants except four for them (see the bottom of [this unicodeset](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3AGrek%3A%5D%26%5B%3ALu%3A%5D%26%5B%3AisNFD%3A%5D&g=&i=)) and look at the characters starting with `0x3Fx`, which doesn't fit in 7 bits as an offset). We could use some of the vowel values as sentinels there.

Also, do we even need to store the uppercase data for consonants? We could fall back to regular uppercasing behavior for those, and simply mark that they _are_ Greek letters in the data. (@eggrobin ?)

Reply:
> Also, do we even need to store the uppercase data for consonants? We could fall back to regular uppercasing behavior for those, and simply mark that they are Greek letters in the data. (@eggrobin ?)

Not for ῥ and ῤ.

Reply:
I went ahead and got rid of this. It's much cleaner, the specialcasing now only applies to vowels and the consonant stuff is only used for "is this a greek letter" checks

Reply:
> Not for ῥ and ῤ.

Speaking of which, do we have tests that would catch that? Otherwise I can add some next week.

Reply:
@eggrobin hmm, the easy solution here is to pretend rho is a vowel, of course

Reply:
> the easy solution here is to pretend rho is a vowel, of course

Not really, that would break other logic. But you could special-case its uppercasing.

Reply:
(we can rename GreekVowel to GreekDiacriticCarrier)

Reply:
@eggrobin which other logic would it break?

Reply:
I guess `preceding_greek_vowel_diacritics()` would change? hm.

Reply:
I could have consonant be `Consonant(is_special_rho)`

Reply:
Done. I could also just special case the two code points in code.